test/sql/index/test_art_wal_replay.cpp

```diff
@@ -0,0 +1,347 @@
+	return *art_ptr;
+}
+} // namespace
+
+// This test inspects the ART tree to check that index WAL operations are properly replayed after
+// restarting the database. This was not being explicitly triggered in any CI or SQL tests, so it is
+// necessary to have a C++ test that explicitly traverses the ART tree to verify it.
+
+// This test includes both physical and generated columns to test that the buffering and column_id mappings
+// work -- it includes a "regular case" with an index on the first two physical columns, and an index on
+// physical columns that are interleaved between  generated columns to test that the buffered mappings work as
+// intended.
+TEST_CASE("Test ART index with WAL replay - generated columns and interleaved inserts/deletes",
+          "[wal][art-wal-replay]") {
+	duckdb::unique_ptr<QueryResult> result;
+	auto db_path = TestCreatePath("art_wal_gen_test.db");
+	DeleteDatabase(db_path);
+	{
+		DuckDB db(nullptr);
+		Connection con(db);
+
+		REQUIRE_NO_FAIL(con.Query("ATTACH '" + db_path + "' AS testdb"));
+		REQUIRE_NO_FAIL(con.Query("USE testdb"));
+		REQUIRE_NO_FAIL(con.Query("PRAGMA disable_checkpoint_on_shutdown"));
+		REQUIRE_NO_FAIL(con.Query("PRAGMA wal_autocheckpoint='1TB'"));
+
+		REQUIRE_NO_FAIL(con.Query("CREATE TABLE tbl(a INT, b INT, c AS (2*a), d VARCHAR, e AS (b + 2), f VARCHAR)"));
+		REQUIRE_NO_FAIL(con.Query("CREATE INDEX idx_ab ON tbl(a, b)"));
+		REQUIRE_NO_FAIL(con.Query("CREATE INDEX idx_df ON tbl(d, f)"));
+
+		REQUIRE_NO_FAIL(
+		    con.Query("INSERT INTO tbl SELECT range, range * 2, 'val_' || range, 'tag_' || range FROM range(100)"));
+
+		REQUIRE_NO_FAIL(con.Query("DELETE FROM tbl WHERE a % 5 = 0"));
+
+		result = con.Query("SELECT COUNT(*) FROM tbl");
+		REQUIRE(CHECK_COLUMN(result, 0, {Value::BIGINT(80)}));
+
+		REQUIRE_NO_FAIL(con.Query("USE memory"));
+		REQUIRE_NO_FAIL(con.Query("DETACH testdb"));
+	}
+
+	{
+		// Reattach database and verify that WAL index replays work.
+		DuckDB db(nullptr);
+		Connection con(db);
+
+		REQUIRE_NO_FAIL(con.Query("ATTACH '" + db_path + "' AS testdb"));
+		REQUIRE_NO_FAIL(con.Query("USE testdb"));
+
+		result = con.Query("SELECT * FROM tbl WHERE a = 1");
+		REQUIRE(CHECK_COLUMN(result, 0, {1}));
+
+		result = con.Query("SELECT COUNT(*) FROM tbl");
+		REQUIRE(CHECK_COLUMN(result, 0, {Value::BIGINT(80)}));
+
+		auto &idx_ab = GetARTIndex(con, "tbl", "idx_ab");
+		auto &idx_df = GetARTIndex(con, "tbl", "idx_df");
+
+		ArenaAllocator arena_ab(BufferAllocator::Get(idx_ab.db));
+		ArenaAllocator arena_df(BufferAllocator::Get(idx_df.db));
+
+		// Verify idx_ab: all mod 5 deleted, everything else exists
+		for (idx_t i = 0; i < 100; i++) {
+			Value val_a(static_cast<int32_t>(i));
+			Value val_b(static_cast<int32_t>(i * 2));
+			auto key = ARTKey::CreateARTKey<int32_t>(arena_ab, val_a);
+			auto key_b = ARTKey::CreateARTKey<int32_t>(arena_ab, val_b);
+			key.Concat(arena_ab, key_b);
+			auto leaf = ARTOperator::Lookup(idx_ab, idx_ab.tree, key, 0);
```

Comment:
Do we need this to be a C++ test? Can't we trigger similar behavior through SQL?

In general we really like to avoid writing C++ tests because it locks us into specific APIs that we might want to change in the future (i.e. more code to refactor if any of the APIs used here ever need to change).

Reply:
Note that our sqllogic test also provides functionality like loops and conditional evaluation, e.g.:

```sql
loop i 0 10

onlyif i=0
statement ok
SELECT ...

endloop
```